### ISSUE TITLE
How to compute the gradient of train loss w.r.t the activation of dense layer?

### ISSUE THREAD

**Issue by lc82111** [Lasagne/Lasagne#523]
I tried to get the gradient by making use theano.grad(train_loss, self.activation), where train_loss is computed by using lasagne.layers.get_output(output_layer),  and to get a hander (Intermediate symbol expression) for the activation, I have  changed the variable "activation" of dense layer to self.activation. But I still get a DisconnectedInputError complaining self.activation not in the ComputeGraph. So what I missed here? Thanks.

**Comment by benanne** [comment]
If you wish to get the gradient of the loss w.r.t. the output of a given layer `l0`, you could do:
```
output, activation = lasagne.layers.get_output([output_layer, l_0])
loss = ...
grad = T.grad(loss, activation)
```

or, perhaps this is cleaner:

```
activation = T.matrix() # or perhaps T.tensor4() or some other dimensionality in your case
output = lasagne.layers.get_output(output_layer, { l0: activation }) # this overrides the output of l0 with the expression you've provided
loss = ...
grad = T.grad(loss, activation)
```

Both untested, but iirc both should work.

By the way, please post questions like this on the mailing list instead: https://groups.google.com/forum/#!forum/lasagne-users
We would like to reserve the GitHub issue queue for development / bug reporting only. Thanks for your cooperation!

**Comment by f0k** [comment]
> Both untested, but iirc both should work.

But note that they do different things. Both get the gradient of the loss wrt. the activation, but the first one does a standard forward pass through the full network (starting from its input layer(s)), while the second one starts the forward pass at layer `l0` with custom input.

**Comment by lc82111** [comment]
thanks, I test the first one, it works. 
And I guess what I need is the standard forward as the f0k said.

**Comment by benanne** [comment]
Good catch.
